### PR TITLE
network: adding some accessors for ALPN work.

### PR DESCRIPTION
**Pull request by alyssawilk**
Adding some basic functionality to the Network::Connection needed to disassociate TCP connections from their connection pool before handing them off to a codec client.

Additional Description:
I really wanted to add Network::Connection Connecting as a state, and I did a PR to that effect and cleaned up the 30 or so call sites in Envoy, but then realized that because it broke nearly every use of state, we either needed to have a separate boolean, or change all the enum values so folks would have to fix their downstream code.  I chose the former as the lesser of two evils.

I also waffled a bit about removing the callbacks or sticking with the remove pattern in codec_helper.h and went with consistency over cleanup.

Risk Level: low (not yet used)
Testing: unit tests
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a
Part of #3431

### CODE REVIEW

File: include/envoy/network/filter.h [hunk 1]
```diff
@@ -227,6 +227,11 @@ class FilterManager {
    */
   virtual void addReadFilter(ReadFilterSharedPtr filter) PURE;
 
+  /**
+   * Remove a read filter from the connection.
+   */
+  virtual void removeReadFilter(ReadFilterSharedPtr filter) PURE;
```
Comment (mattklein123) [inline include/envoy/network/filter.h:233]: What is the use case for this one? Naively I would assume that we can just make a raw connection without any read filters, detect ALPN, and then attach an HTTP filter, etc. and continue?

**Review by mattklein123 — COMMENTED** [review]
Agreed on the tradeoffs around state, oh well. One question about the interface change and also needs a master merge.

/wait

File: include/envoy/network/filter.h (hunk 1 above)
Comment (alyssawilk) [reply include/envoy/network/filter.h:233]: the tcp pool by default adds a read filter, and disconnects if data is read when no session is associated.
We could skip that and the add-and-remove but I'd want something to make sure early data didn't go into the void, so I'm mildly inclined to leave as-is.

Comment (mattklein123) [reply include/envoy/network/filter.h:233]: Ah OK. I see. I guess it would be nice to avoid this interface change but up to you.

**Review by mattklein123 — APPROVED** [review]
LGTM assuming you think these are all needed. Do you want to land this or review the entire change first to see what we think about all of it together?

**Comment by alyssawilk** [comment]
yeah, I've got a PR which works (with hard-coded ALPN) for either HTTP/1 or HTTP/2 integration tests, so I'm pretty solid these will land (and I'll back them out if for some reason the plan drastically changes)
Trying to split smaller things out to make the core code easier to review

**Pull request merged**